Feature/availability atom

Adamik10 opened the pull request:
#### What does this PR do?
- This PR introduces the Availability label component
- It also adds new functionality to the pagefold atom 
     - option to specify color, and size of the pagefold

#### Where should the reviewer start?


#### How can the reviewer see this in action?
- by running "npm run dev", and checking the component out in the browser - under Components/ Availability label

#### Any background context you want to provide?
No.

#### Is this related to processing of personal data?
No.

#### What are the relevant tickets?
N/A

#### Screenshots (if appropriate)
N/A

#### Questions for the reviewer:
What do you think of the implementation?

Adamik10 committed:
Updated pagefold atom with new features
- custom color
- possible to remove the triangle fold
- updated spacing with an extra small value for pagefold

Adamik10 committed:
Created Availability Label component
- created folder structure for the component with a .tsx, stories.tsx, and .scss file
- updated typography to be used in the Availability label
- added comments to base.scss and AL import

Adamik10 committed:
Minor fixes to Availability label component
- amended the scss for availability label, in order to comply with the existing repository standards
- this also means that we had to adjust the .tsx and story.tsx files

Adamik10 committed:
Deleting .vscode folder, and adding a .vscode.example for vscode settings
- the .vscode folder was already in .gitignore, but started being tracked on accident
- we introduce a suggestion for what an example folder should look like in form of .vscode.example

kasperg commented on src/stories/availability-label/AvailabilityLabel.tsx:
```diff
@@ -0,0 +1,50 @@
+import { Pagefold, PagefoldProps } from "../pagefold/Pagefold";
+
+export type AvailabilityLabelPropsType = {
+  manifestation: "bog" | "ebog" | "lydbog (net)" | "lydbog (cd-mp3)";
+  availability: "Hjemme" | "Online" | "Udlånt";
```
Is there a reason behind the capitalization here? I mean "bog" is all lower case while "Hjemme" has initial upper case.

In general I would expect text to have a natural casing (here initial upper case) and if special casing was required by the design then we would modify it using CSS.

Adamik10 replied:
In regards to the capitalization - I wasn't sure what the data that would look like when we get if from the API, so I just made it random and not the same. 😅 I'll change it to be consistent 🤝

kasperg commented on src/stories/availability-label/AvailabilityLabel.stories.tsx:
```diff
@@ -0,0 +1,56 @@
+  pagefoldProps: {
+    inherit: false,
+    container: false,
+    size: "xsmall",
+    classes: "availability-label availability-label--unselected text-label",
+    colorClass: "success",
```
These seem to be duplicated from `getPageFoldProps()` where they are derived from `status`. Do they have to be defined as a part of the story?

The same goes for the next two stories.

Adamik10 replied:
You are completely right! We don't actually need these.

kasperg commented on src/stories/pagefold/pagefold.scss:
```diff
@@ -42,5 +46,13 @@ $pagefold-map: (
     top: 0;
     left: 0;
+
+    &.success {
+      background-color: $c-signal-success;
+    }
+
+    &.alert {
+      background-color: $c-signal-alert;
+    }
```
If we are to follow the BEM structure of the project and reduce nesting I would suggest that we change this to:

```scss
.pagefold-triangle--success {
  background-color: $c-signal-success;
}

.pagefold-triangle--alert {
  background-color: $c-signal-alert;
}
```

Adamik10 replied:
This makes total sense. A class that is just called .success is too generic. I fixed it 😸

kasperg commented on src/stories/pagefold/pagefold.scss:
```diff
@@ -16,9 +18,11 @@ $pagefold-map: (
+    min-width: #{$triangle-width}px;
+    min-height: #{$triangle-width}px;
     position: relative;
+    background-color: #eee9e5;
```
Do we have a variable for this color?

Adamik10 replied:
Yeps! Adding it.

kasperg commented on src/stories/typography/typography.scss:
```diff
@@ -318,8 +318,17 @@
+
+.text-label-normal {
+  @extend %text-label;
+  font-weight: 400;
```
To address this and keep our BEM structure I suggest simplifying this and requiring users to add both classes to their elements.

```scss
.text-label--normal {
  font-weight: 400;
}
```

Also consider adding stories to these variations.

Adamik10 replied:
Fixed, and done 😄

kasperg replied:
Remember to use double dashes `--` between the element `text-label` and the modifier `normal`

Adamik10 replied:
mmmm we will need to talk about the conventions. 
I agree and would love to be using this, but it is unfortunately not the current reality in the repository.

kasperg reviewed (changes requested):
:+1: Nice things here. I have a few comments that I would like you to look into.

spaceo commented on base.scss:
```diff
@@ -1,6 +1,8 @@
 @import "./src/stories/colors/color-variables";
 
+// the imports below are not advised to be moved around, because
+// they cross-reference various defined variables between each other
```
👍 Good comment. I think you should start the sentence with a capital letter. We do that everywhere else.

Adamik10 replied:
Done 🥇  Thanks for pointing it out.

spaceo commented on src/stories/pagefold/Pagefold.tsx:
```diff
@@ -1,22 +1,31 @@
+  size: "none" | "xsmall" | "small" | "medium" | "large" | "xlarge";
+  colorClass?: "success" | "alert";
   children?: React.ReactNode;
   classes?: string;
```
In React it is a convention to use the prop classNames for classes. I think you should not deviate from that.

Adamik10 commented on src/stories/typography/typography.scss:
```diff
@@ -318,8 +318,17 @@
+
+.text-label-semibold {
+  @extend %text-label;
+  font-weight: 500;
```
To address this and keep our BEM structure I simplified this and require users to add both classes to their elements.

spaceo committed:
Cleaning up class construction and remove uneeded pagefoldProps
* Removing pagefold props from availability label component
  We know what we want to pass inside of the availability component
  so we don't need to expose it
* Make use of map object to construct pagefold props
* Introducing clsx package for cleaner className construction
* Move pagefold prop logic from availability label component to HOC

Adamik10 committed:
Bug fixes to Pagefold.stories.tsx after refactoring

Adamik10 committed:
Merge pull request #4 from reload/feature/availability-atom_more-declarative-and-with-clsx

Cleaning up class construction and remove uneeded pagefoldProps
- Removing pagefold props from availability label component
- We know what we want to pass inside of the availability component
so we don't need to expose it
- Make use of map object to construct pagefold props
- Introducing clsx package for cleaner className construction
- Move pagefold prop logic from availability label component to HOC

Adamik10 committed:
Optimilize and bug-fix the Availability label based on code review

Bug fixes and code refactoring solves folloing issues:

- Documents the base scss, so that developers don't change order of
imports, which can break the code due to variable declaration.

- Availability label higher order component was returning incorrect
status classses, which caused the AL component not to change color.

- Availability label story arguments had inconsistent capitalization.

- Pagefold style  was using a color hex code instead of a defined
variable.

- Pagefold style was using generic short classes, which did not follow
the BEM structure of the project, and targeted unwanted components as
well.

- Pagefold component type prop was wrongly defined.

- The typogtraphy was using generic, short class names, which did not
follow the BEM structure of the project, and could target unwanted
components as well.

- Extra typography classes were added, but no stories were made for
them.

Adamik10 committed:
Avoid using math.random in the Material component

We used math.random to generate a color tint for the placeholder
material. As a result, the github action UI tests would consider a
different color tint class as a change to the UI that we then manually
need to approve.

This is avoided by using storybook controls, where the tint can be
adjusted in the story at runtime.

spaceo commented on src/stories/pagefold/Pagefold.tsx:
```diff
@@ -5,7 +5,7 @@ export type PagefoldProps = {
   isAContainer: boolean;
   size: "none" | "xsmall" | "small" | "medium" | "large" | "xlarge";
-  type?: "success" | "alert";
+  type?: string;
```
why not being more strict here and specify known types?

Adamik10 replied:
I fixed this to work in a more declarative way :)

spaceo commented on src/stories/pagefold/Pagefold.tsx:
```diff
@@ -25,15 +25,13 @@ export const Pagefold = (props: PagefoldProps) => {
-      {"pagefold-inherit-parent": isInheriting}
-    )
-  }
+    content: clsx(type, `pagefold-triangle--${size}`, {
```
Honestly I find the former syntax more readable. But don't want to be square about it.

Adamik10 replied:
Oh yes, I agree. 
We should definitely find a way to let the editor correct these things. 
Otherwise I'll just need to watch out for it. 😭

spaceo commented on src/stories/material/Material.tsx:
```diff
@@ -1,28 +1,35 @@
 export const Material = (props: MaterialProps) => {
-  // If Material does not have an img, show a random bg-color for each card.
-  const bg = placeholderBg[(Math.random() * placeholderBg.length) | 0];
+  const { size, animate, url } = props;
```
Why not extract tint here as well?:
```
  const { size, animate, url, tint } = props;
```

Adamik10 replied:
Good point. I was a bit confused over whether it will always be defined or not. But in either case, you're right.

spaceo commented on src/stories/material/Material.tsx:
```diff
@@ -1,28 +1,35 @@
-  const bg = placeholderBg[(Math.random() * placeholderBg.length) | 0];
+  const { size, animate, url } = props;
+
+  const classes = {
+    wrapper: clsx(
+      `material material--${size}`,
+      { "bg-identity-tint-120": props.tint === "120" },
+      { "bg-identity-tint-80": props.tint === "80" },
+      { "bg-identity-tint-60": props.tint === "60" },
+      { "bg-identity-tint-40": props.tint === "40" },
+      { "bg-identity-tint-20": props.tint === "20" },
+      { material__animate: animate }
+    ),
+  };
+
```
Could be simplified, I think. (Not tested):
```

const placeholderBg = {
  "120": "bg-identity-tint-120",
  "80": "bg-identity-tint-80",
  "60": "bg-identity-tint-60",
  "40": "bg-identity-tint-40",
  "20": "bg-identity-tint-20"
};

const classes = {
  wrapper: clsx(
    `material material--${size}`,
    placeholderBg[tint] || false,
    { material__animate: animate }
  ),
};
```

Adamik10 replied:
That is very true :) optimized it.

spaceo commented on src/stories/material/Material.tsx:
```diff
@@ -1,28 +1,35 @@
+        <img
+          src={url}
+          alt=""
+          onError={(e) => (e.currentTarget.style["display"] = "none")}
```
I was not aware of this construction the first time I reviewed.
Just curious: What is happening here? If an image cannot be displayed it gets `style="display: none"` and is hidden?

kasperg replied:
This relates to work done by @PELN in [57920b9](https://github.com/reload/dpl-design-system/commit/57920b9350931f8ce1d1c94524b9f7a06e0c7e8f)

She writes: "Add a function to handle broken image icon which should not be displayed"

Perhaps she can provide more details. It is not related to the current PR anyway.

Adamik10 replied:
I am resolving this, as it is not related to the PR.

Adamik10 committed:
Optimize pagefold "type" prop and its usage for more declarative code

The "type" prop now accepts only two specific values. The component then
decides which classes to apply to itself based on the prop.

This avoids using imperative code, where we would have to pass down very
specific class names.

Adamik10 committed:
Define tint classes in a separate constant for "Material" component

Increases code readability in the material component.

Adamik10 committed:
Use @extend for text label classes

This utilizes the text label placeholder, which lets us avoid duplicate
statements.

Adamik10 committed:
Comply with the prettier and eslint setup for Availability label HOC

Adjusts formatting of the Availability label Higher order component,
which should always be followed as a common standard.

Adamik10 commented:
Can we somehow fix the github action check fails @kasperg @spaceo ?

Adamik10 committed:
Comply with the prettier and eslint setup for Material Banner component

Adjusts formatting of the Material Banner  component, which should
always be followed as an established common standard.

Adamik10 committed:
Do not require the tint parameter in MaterialCard component

After recently adding tint as a property in Material component,
MaterialCard component, which uses Material, made all its props
required. This caused Typescript compile errors.

We now don'r require the tint prop.

Adamik10 committed:
Use tint prop when displaying multiple Material components

We added tint prop in Material component, which was already used
multiple places around the app. When using multiple instances of this
component in succession, the color tint was set to a default value, and
didn't follow the design standards established in figma.

We now specify the tints for this use case.

Adamik10 committed:
Remove unused imports in Modal Details component

Adamik10 committed:
Merge branch 'main' into feature/availability-atom

Resolves the merge conflicts for this pull request.

spaceo reviewed (approved):
Very nice!

kasperg reviewed (approved):
:+1: LGTM

Adamik10 committed:
Do not specify background color for pagefold component

Components that were already using pagefold were unintentionally
affected by adding specific background color and display property.
We now don't specify background color, and inherit the display prop.
